File: imageCropper/src/main/java/com/crust87/imagecropper/cropbox/CropBox.java
/*
 * ImageCropper
 * https://github.com/crust87/Android-ImageCropper
 *
 * Mabi
 * crust87@gmail.com
 * last modify 2015-05-19
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.crust87.imagecropper.cropbox;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.Rect;
import android.view.MotionEvent;

import com.crust87.imagecropper.ImageCropper;
import com.crust87.imagecropper.R;

public abstract class CropBox {
	public enum ACTION_LIST{resize, move, none}

	// Components
	protected Context mContext;
	protected Paint mPaint;
	protected Paint mMaskPaint1;
	protected Paint mMaskPaint2;
	protected Paint mBitmapPaint;
	protected Bitmap mBitmap;
	protected Canvas mCanvas;

	// Attributes
	protected Rect mBound;
	protected int mMinSize;
	protected int mMinHalfSize;
	protected int mDefaultHalfSize;
	protected float mX;
	protected float mY;
	protected float mPostX;
	protected float mPostY;
	protected float mScale;
	protected int mBoxColor;
	protected int mLineWidth;
	protected int mAnchorSize;

	// Working Variables
	protected ACTION_LIST mCurrentEvent;

	protected CropBox(Context context) {
		mContext = context;

		mMinSize = mContext.getResources().getDimensionPixelSize(R.dimen.min_box_Size);
		mMinHalfSize = mMinSize / 2;
		mDefaultHalfSize = mContext.getResources().getDimensionPixelSize(R.dimen.default_box_size) / 2;
	}

	protected void setAttributes(float leftMargin, float topMargin, Rect bound, float scale, int boxColor, int lineWidth, int anchorSize) {
		mX = leftMargin + bound.width()/2;
		mY = topMargin + bound.height()/2;
		mBound = bound;
		mScale = scale;
		mBoxColor = boxColor;
		mLineWidth = lineWidth;
		mAnchorSize = anchorSize;
	}

	protected void init() {
		mPaint = new Paint();
		mPaint.setColor(Color.WHITE);
		mPaint.setAntiAlias(true);
		mPaint.setStyle(Paint.Style.STROKE);
		mPaint.setStrokeWidth(mLineWidth);
		mPaint.setColor(mBoxColor);
		mPaint.setStrokeWidth(mLineWidth);

		mMaskPaint1 = new Paint();
		mMaskPaint1.setColor(Color.BLACK);
		mMaskPaint1.setAntiAlias(true);

		mMaskPaint2 = new Paint();
		mMaskPaint2.setColor(Color.WHITE);
		mMaskPaint2.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.XOR));

		mBitmapPaint = new Paint();
		mBitmapPaint.setFilterBitmap(true);
		mBitmapPaint.setAlpha(128);

		mBitmap = Bitmap.createBitmap(mBound.width(), mBound.height(), Bitmap.Config.ARGB_8888);
		mCanvas = new Canvas(mBitmap);
	}

	public abstract boolean processTouchEvent(MotionEvent event);
	public boolean move(float x, float y) {
		mX -= x;
		mY -= y;

		return true;
	}

	protected abstract void setAnchor();

	public abstract boolean contains(float x, float y);
	public abstract void draw(Canvas canvas);

	protected abstract float getX();
	protected abstract float getY();
	protected abstract float getWidth();
	protected abstract float getHeight();

	public abstract int getCropX();
	public abstract int getCropY();
	public abstract int getCropWidth();
	public abstract int getCropHeight();

	@Override
	public String toString() {
		return "view x: " + getX() + " y: " + getY() + " width " + getWidth();
	}

	public void setColor(int color) {
		mBoxColor = color;
		mPaint.setColor(mBoxColor);
	}

	public void setColor(String colorCode) {
		setColor(Color.parseColor(colorCode));
	}

	public int getLineWidth() {
		return mLineWidth;
	}

	public void setLineWidth(int lineWidth) {
		mLineWidth = lineWidth;
		mPaint.setStrokeWidth(mLineWidth);
	}

	public int getAnchorSize() {
		return mAnchorSize;
	}

	public void setAnchorSize(int anchorSize) {
		mAnchorSize = anchorSize;
		onSetAnchorSize();
	}

	protected abstract void onSetAnchorSize();

	public static class CropBoxBuilder {
		private int newBoxType;
		private float newLeftMargin;
		private float newTopMargin;
		private Rect newBound;
		private float newScale;
		private int newBoxColor;
		private int newLineWidth;
		private int newAnchorSize;

		public CropBoxBuilder setBoxType(int boxType) {
			newBoxType = boxType;
			return this;
		}

		public CropBoxBuilder setLeftMargin(float leftMargin) {
			newLeftMargin = leftMargin;
			return this;
		}

		public CropBoxBuilder setTopMargin(float topMargin) {
			newTopMargin = topMargin;
			return this;
		}

		public CropBoxBuilder setBound(Rect bound) {
			newBound = bound;
			return this;
		}

		public CropBoxBuilder setScale(float scale) {
			newScale = scale;
			return this;
		}

		public CropBoxBuilder setBoxColor(int boxColor) {
			newBoxColor = boxColor;
			return this;
		}

		public CropBoxBuilder setLineWidth(int lineWidth) {
			newLineWidth = lineWidth;
			return this;
		}

		public CropBoxBuilder setAnchorSize(int anchorSize) {
			newAnchorSize = anchorSize;
			return this;
		}

		public CropBox createCropBox(Context context) {
			CropBox cropBox;
			switch (newBoxType) {
				case ImageCropper.CIRCLE_CROP_BOX:
					cropBox = new CircleCropBox(context);
					break;
				case ImageCropper.RECT_CROP_BOX:
					cropBox = new RectCropBox(context);
					break;
				default:
					cropBox = new CircleCropBox(context);
					break;
			}

			cropBox.setAttributes(newLeftMargin, newTopMargin, newBound, newScale, newBoxColor, newLineWidth, newAnchorSize);
			cropBox.init();

			return cropBox;
		}
	}
}

File: imageCropper/src/main/java/com/crust87/imagecropper/cropbox/CropBox.kt
/*
 * ImageCropper
 * https://github.com/crust87/Android-ImageCropper
 *
 * Mabi
 * crust87@gmail.com
 * last modify 2019-01-19
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.crust87.imagecropper.cropbox

import android.content.Context
import android.graphics.*
import android.view.MotionEvent
import com.crust87.imagecropper.R
import com.crust87.imagecropper.cropbox.anchor.Anchor

class CropBox(context: Context, val leftMargin: Float, val topMargin: Float, val bound: Rect, var boxColor: Int, val lineWidth: Int, val anchorSize: Int) {

    companion object {
        const val TOP_LEFT = 0
        const val TOP_RIGHT = 1
        const val BOTTOM_LEFT = 2
        const val BOTTOM_RIGHT = 3
        val ANCHOR_LIST = arrayListOf(TOP_LEFT, TOP_RIGHT, BOTTOM_LEFT, BOTTOM_RIGHT)
    }

    enum class Action {
        Resize, Move, None
    }

    val minSize = context.resources.getDimensionPixelSize(R.dimen.min_box_Size).toFloat()

    val paint = Paint().apply {
        color = Color.WHITE
        isAntiAlias = true
        style = Paint.Style.STROKE
        strokeWidth = lineWidth.toFloat()
        color = boxColor
        strokeWidth = lineWidth.toFloat()
    }

    val backgroundPaint = Paint().apply {
        color = Color.RED
        isAntiAlias = true
    }

    val holePaint = Paint().apply {
        color = Color.WHITE
        xfermode = PorterDuffXfermode(PorterDuff.Mode.XOR)
    }

    val maskPaint = Paint().apply {
        isFilterBitmap = true
        alpha = 128
    }

    val anchors = ArrayList<Anchor>()

    /*
    Attributes
     */
    var x = 0f
    var y = 0f
    var width = minSize
    var height = minSize

    /*
    Working Variables
     */
    private var maskBitmap: Bitmap
    private var maskCanvas: Canvas
    private var postX = 0f
    private var postY = 0f
    private var currentEvent: Action = Action.None
    private var currentAnchor = -1

    init {
        maskBitmap = Bitmap.createBitmap(bound.width(), bound.height(), Bitmap.Config.ARGB_8888)
        maskCanvas = Canvas(maskBitmap)

        initAnchor()
        setAnchor()
    }

    fun processTouchEvent(event: MotionEvent): Boolean {
        val eventX = event.x - leftMargin
        val eventY = event.y - topMargin

        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                postX = eventX
                postY = eventY

                for (anchor in anchors) {
                    if (anchor.contains(eventX, eventY)) {
                        currentEvent = Action.Resize
                        currentAnchor = anchor.id

                        return false
                    }
                }

                currentEvent = when (contains(eventX, eventY)) {
                    true -> Action.Move
                    false -> Action.None
                }

                return false
            }
            MotionEvent.ACTION_MOVE -> {
                val dx = postX - eventX
                val dy = postY - eventY

                when (currentEvent) {
                    Action.Move -> move(dx, dy)
                    Action.Resize -> resize(dx, dy)
                }

                setAnchor()

                postX = eventX
                postY = eventY

                return true
            }

            MotionEvent.ACTION_UP -> {
                currentEvent = Action.None

                return false
            }
        }

        return false
    }

    fun move(dx: Float, dy: Float): Boolean {
        var deltaX = dx
        var deltaY = dy

        val top = y - dy
        if (top < (bound.top - topMargin)) {
            deltaY = dy - (bound.top - topMargin) + top
        }

        val bottom = y + height - dy
        if (bottom > (bound.bottom - topMargin)) {
            deltaY = dy - (bound.bottom - topMargin) + bottom
        }

        val left = x - dx
        if (left < (bound.left - leftMargin)) {
            deltaX = dx - (bound.left - leftMargin) + left
        }

        val right = x + width - dx
        if (right > (bound.right - leftMargin)) {
            deltaX = dx - (bound.right - leftMargin) + right
        }

        x -= deltaX
        y -= deltaY

        return true
    }

    fun resize(dx: Float, dy: Float): Boolean {
        if (currentAnchor != -1) {
            var deltaX = x
            var deltaY = y
            var deltaWidth = width
            var deltaHeight = height

            when (currentAnchor) {
                TOP_LEFT -> {
                    deltaX -= dx
                    deltaY -= dy
                    deltaWidth += dx
                    deltaHeight += dy
                }
                TOP_RIGHT -> {
                    deltaY -= dy
                    deltaWidth -= dx
                    deltaHeight += dy
                }
                BOTTOM_LEFT -> {
                    deltaX -= dx
                    deltaWidth += dx
                    deltaHeight -= dy
                }
                BOTTOM_RIGHT -> {
                    deltaWidth -= dx
                    deltaHeight -= dy
                }
            }

            if (deltaX < bound.left - leftMargin) {
                deltaX = bound.left - leftMargin
                deltaWidth = (x + width) - deltaX
            }

            if (deltaX + deltaWidth > bound.right - leftMargin) {
                deltaWidth = bound.right - leftMargin - deltaX
            }

            if (deltaY < bound.top - topMargin) {
                deltaY = bound.top - topMargin
                deltaHeight = (y + height) - deltaY
            }

            if (deltaY + deltaHeight > bound.bottom - topMargin) {
                deltaHeight = bound.bottom - topMargin - deltaY
            }

            if (deltaWidth < minSize) {
                if (currentAnchor == TOP_LEFT || currentAnchor == BOTTOM_LEFT) {
                    deltaX = (x + width) - minSize
                }

                deltaWidth = minSize
            }

            if (deltaHeight < minSize) {
                if (currentAnchor == TOP_LEFT || currentAnchor == TOP_RIGHT) {
                    deltaY = (y + height) - minSize
                }

                deltaHeight = minSize
            }

            x = deltaX
            y = deltaY
            width = deltaWidth
            height = deltaHeight

            return true
        }

        return false
    }

    fun contains(ex: Float, ey: Float): Boolean {
        return ex >= x && ex <= x + width && ey >= y && ey <= y + height
    }

    open fun initAnchor() = ANCHOR_LIST.map {
        anchors.add(Anchor(it, anchorSize / 2).apply {
            setColor(boxColor)
        })
    }

    open fun setAnchor() = anchors.map { anchor ->
        when (anchor.id) {
            TOP_LEFT -> anchor.setLocation(x, y)
            TOP_RIGHT -> anchor.setLocation(x + width, y)
            BOTTOM_LEFT -> anchor.setLocation(x, y + height)
            BOTTOM_RIGHT -> anchor.setLocation(x + width, y + height);
        }
    }

    fun draw(canvas: Canvas) {
        maskCanvas.drawRect(0f, 0f, canvas.width.toFloat(), canvas.height.toFloat(), backgroundPaint)
        maskCanvas.drawRect(x, y, x + width, y + height, holePaint)
        canvas.drawBitmap(maskBitmap, null, bound, maskPaint)

        canvas.translate(leftMargin, topMargin)
        canvas.drawRect(x, y, x + width, y + height, paint)
        if (currentEvent != Action.Move) {
            anchors.map { anchor ->
                anchor.draw(canvas)
            }
        }
    }

    override fun toString(): String {
        return "view x: $x, y: $y, width: $width, height: $height"
    }

    class CropBoxBuilder {
        private var newBoxType: Int = 0
        private var newLeftMargin: Float = 0.toFloat()
        private var newTopMargin: Float = 0.toFloat()
        private var newBound: Rect = Rect()
        private var newBoxColor: Int = 0
        private var newLineWidth: Int = 0
        private var newAnchorSize: Int = 0

        fun setBoxType(boxType: Int): CropBoxBuilder {
            newBoxType = boxType
            return this
        }

        fun setLeftMargin(leftMargin: Float): CropBoxBuilder {
            newLeftMargin = leftMargin
            return this
        }

        fun setTopMargin(topMargin: Float): CropBoxBuilder {
            newTopMargin = topMargin
            return this
        }

        fun setBound(bound: Rect): CropBoxBuilder {
            newBound = bound
            return this
        }

        fun setBoxColor(boxColor: Int): CropBoxBuilder {
            newBoxColor = boxColor
            return this
        }

        fun setLineWidth(lineWidth: Int): CropBoxBuilder {
            newLineWidth = lineWidth
            return this
        }

        fun setAnchorSize(anchorSize: Int): CropBoxBuilder {
            newAnchorSize = anchorSize
            return this
        }

        fun createCropBox(context: Context): CropBox {
            return CropBox(context, newLeftMargin, newTopMargin, newBound, newBoxColor, newLineWidth, newAnchorSize)
        }
    }
}
